packages/SwingSet/src/liveslots/watchedPromises.js

```diff
@@ -0,0 +1,186 @@
+      syscall.vatstoreSet('watcherTableID', watcherTableID);
+      // artificially increment the table's refcount so it never gets GC'd
+      vrm.addReachableVref(watcherTableID);
+    }
+
+    let watchedPromiseTableID = syscall.vatstoreGet('watchedPromiseTableID');
+    if (watchedPromiseTableID) {
+      watchedPromiseTable = convertSlotToVal(watchedPromiseTableID);
+    } else {
+      watchedPromiseTable = makeScalarBigMapStore('watchedPromises', {
+        keySchema: M.string(),
+        durable: true,
+      });
+      watchedPromiseTableID = convertValToSlot(watchedPromiseTable);
+      syscall.vatstoreSet('watchedPromiseTableID', watchedPromiseTableID);
+      // artificially increment the table's refcount so it never gets GC'd
+      vrm.addReachableVref(watchedPromiseTableID);
+    }
+  }
+
+  function watchPromise(p, watcher, ...args) {
+    const watcherVref = convertValToSlot(watcher);
+    assert(watcherVref, 'invalid watcher');
+    const { virtual } = parseVatSlot(watcherVref);
+    assert(virtual, 'promise watcher must be a virtual object');
+    if (watcher.onFulfilled) {
+      assert.typeof(watcher.onFulfilled, 'function');
+    }
+    if (watcher.onRejected) {
+      assert.typeof(watcher.onRejected, 'function');
+    }
+    assert(
+      watcher.onFulfilled || watcher.onRejected,
+      'promise watcher must implement at least one handler method',
+    );
+
+    const vpid = convertValToSlot(p);
+    assert(vpid, 'invalid promise');
+    const { type } = parseVatSlot(vpid);
+    assert(type === 'promise', 'watchPromise only watches promises');
+    if (watchedPromiseTable.has(vpid)) {
+      const watches = watchedPromiseTable.get(vpid);
+      watchedPromiseTable.set(vpid, harden([...watches, [watcher, ...args]]));
+    } else {
+      watchedPromiseTable.init(vpid, harden([[watcher, ...args]]));
+      pseudoThen(p, vpid);
+    }
+  }
+
+  function providePromiseWatcher(
+    kindHandle,
+    fulfillHandler = x => x,
+    rejectHandler = x => {
+      throw x;
+    },
+  ) {
+    assert.typeof(fulfillHandler, 'function');
+    assert.typeof(rejectHandler, 'function');
+
+    const makeWatcher = defineDurableKind(kindHandle, () => ({}), {
+      // @ts-ignore  TS is confused by the spread operator
+      onFulfilled: (_context, res, ...args) => fulfillHandler(res, ...args),
+      // @ts-ignore
+      onRejected: (_context, rej, ...args) => rejectHandler(rej, ...args),
+    });
+
+    if (promiseWatcherByKindTable.has(kindHandle)) {
+      return promiseWatcherByKindTable.get(kindHandle);
+    } else {
+      const watcher = makeWatcher();
+      promiseWatcherByKindTable.init(kindHandle, watcher);
+      return watcher;
+    }
+  }
+
+  function prepareShutdownRejections(deciderVPIDs) {
+    const deadPromises = [];
+    for (const vpid of deciderVPIDs) {
+      if (watchedPromiseTable.has(vpid)) {
+        deadPromises.push(vpid);
+      }
+    }
+    syscall.vatstoreSet('deadPromises', deadPromises.join(','));
```

Comment:
the use of `,` as a delimiter means we should probably update the comment in `parseVatSlot.js` to prohibit comma as a component in vrefs, with a pointer to this file and the `deadPromises` record